### PR TITLE
[MI-2216] Pull request link preview and updated UI for task link preview

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
Screenshot from 2022-10-12 16-32-57

### CODE REVIEW

File: server/plugin/client.go [hunk 1]
```diff
@@ -88,7 +89,7 @@ func (c *client) CreateTask(body *serializers.CreateTaskRequestPayload, mattermo
 }
 
 // Function to get the task.
-func (c *client) GetTask(organization, taskID, mattermostUserID string) (*serializers.TaskValue, int, error) {
+func (c *client) GetTask(organization, taskID, projectName, mattermostUserID string) (*serializers.TaskValue, int, error) {
```
Comment (ayusht2810) [inline server/plugin/client.go:92]: remove `projectName` if it is not used anywhere.

File: server/plugin/client.go [hunk 2]
```diff
@@ -100,6 +101,19 @@ func (c *client) GetTask(organization, taskID, mattermostUserID string) (*serial
 	return task, statusCode, nil
 }
 
+// Function to get the pull request.
+func (c *client) GetPullRequest(organization, pullRequestID, projectName, mattermostUserID string) (*serializers.PullRequest, int, error) {
```
Comment (ayusht2810) [inline server/plugin/client.go:105]: same

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -99,18 +100,25 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
...
 		newPost, msg := p.PostTaskPreview(taskData, post.UserId, post.ChannelId)
 		return newPost, msg
 	}
+
+	// Check if message contains a pull request link.
+	if pullRequestData, link, isValid := isValidPullRequestLink(post.Message); isValid {
+		newPost, msg := p.PostPullRequestPreview(pullRequestData, link, post.UserId, post.ChannelId)
+		return newPost, msg
+	}
 	return nil, ""
```
Comment (ayusht2810) [inline server/plugin/plugin.go:114]:
```suggestion
	}

	return nil, ""
```

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -119,3 +127,17 @@ func isValidTaskLink(msg string) ([]string, bool) {
...
+	pullRequestRegex := regexp.MustCompile(constants.PullRequestRegex)
+	link := pullRequestRegex.FindString(msg)
+	data := strings.Split(link, "/")
+	if len(data) != 9 {
+		return nil, "", false
+	}
+	if (data[0] != constants.HTTPS && data[0] != constants.HTTP) || data[2] != constants.AzureDevopsBaseURL || data[5] != constants.Git || data[7] != constants.PullRequest {
+		return nil, "", false
+	}
+	return data, link, true
```
Comment (ayusht2810) [inline server/plugin/plugin.go:142]:
```suggestion
	if len(data) != 9 {
		return nil, "", false
	}

	if (data[0] != constants.HTTPS && data[0] != constants.HTTP) || data[2] != constants.AzureDevopsBaseURL || data[5] != constants.Git || data[7] != constants.PullRequest {
		return nil, "", false
	}

	return data, link, true
```

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -119,3 +127,17 @@ func isValidTaskLink(msg string) ([]string, bool) {
 	}
 	return data, true
 }
+
+// Function to validate the pull request link.
+func isValidPullRequestLink(msg string) ([]string, string, bool) {
+	pullRequestRegex := regexp.MustCompile(constants.PullRequestRegex)
+	link := pullRequestRegex.FindString(msg)
+	data := strings.Split(link, "/")
```
Comment (ayusht2810) [inline server/plugin/plugin.go:135]: this is common logic used in the above function as well. make a function for this, if possible

File: server/plugin/taskLinkPreview.go [hunk 6]
```diff
@@ -25,14 +26,78 @@ func (p *Plugin) PostTaskPreview(linkData []string, userID, channelID string) (*
...
+				Title: "Description",
+				Value: description,
+			},
+		},
+		Footer:     linkData[4],
+		FooterIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.ProjectIcon),
+	}
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
+
+	return post, ""
```
Comment (ayusht2810) [inline server/plugin/taskLinkPreview.go:57]:
```suggestion
	}

	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
	return post, ""
```

File: server/plugin/plugin.go [hunk 7]
```diff
@@ -119,3 +127,17 @@ func isValidTaskLink(msg string) ([]string, bool) {
...
+	if len(data) != 9 {
+		return nil, "", false
+	}
+	if (data[0] != constants.HTTPS && data[0] != constants.HTTP) || data[2] != constants.AzureDevopsBaseURL || data[5] != constants.Git || data[7] != constants.PullRequest {
```
Comment (ayusht2810) [inline server/plugin/plugin.go:139]: we are checking the URL by comparing it with regex. do we need to explicitly check it like this?

File: server/plugin/plugin.go [hunk 8]
```diff
@@ -101,43 +101,28 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
...
-	taskRegex := regexp.MustCompile(constants.TaskRegex)
-	link := taskRegex.FindString(msg)
-	data := strings.Split(link, "/")
-	if len(data) != 8 {
-		return nil, false
-	}
-	if (data[0] != constants.HTTPS && data[0] != constants.HTTP) || data[2] != constants.AzureDevopsBaseURL || data[5] != constants.Workitems || data[6] != constants.Edit {
-		return nil, false
+// Function to check if link is present in the message and return link data
+func IsLinkPresent(msg string, regex string) ([]string, string, bool) {
```
Comment (ayusht2810) [inline server/plugin/plugin.go:119]: move this function above the function `MessageWillBePosted` as it is used before.

File: server/plugin/plugin_test.go [hunk 9]
```diff
@@ -36,40 +37,35 @@ func TestMessageWillBePosted(t *testing.T) {
...
+		data        []string
+		isValidLink bool
+		link        string
 	}{
 		{
-			description:     "test change post for valid link",
-			taskData:        []string{"https:", "", "dev.azure.com", "abc", "xyz", "_workitems", "edit", "1"},
-			message:         "mockMessage",
-			isValidTaskLink: true,
+			description: "test change post for valid link",
```
Comment (ayusht2810) [inline server/plugin/plugin_test.go:47]:
```suggestion
			description: "MessageWillBePosted: test change post for valid link",
```

Comment (ayusht2810) [reply server/plugin/plugin_test.go:47]: update everywhere

**Review by ayusht2810 — APPROVED** [review]

File: server/constants/constants.go [hunk 10]
```diff
@@ -40,6 +40,12 @@ const (
 	AzureDevopsBaseURL = "dev.azure.com"
 	Workitems          = "_workitems"
 	Edit               = "edit"
+	TaskLinkRegex      = `http(s)?:\/\/dev.azure.com\/[a-zA-Z0-9!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]*\/[a-zA-Z0-9!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]*\/_workitems\/edit\/[0-9]*`
+
+	// Get pull request link preview constants
+	PullRequest          = "pullrequest"
+	Git                  = "_git"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:47]: is this a prefix if it is then we can include this in its name

File: server/plugin/taskLinkPreview.go [hunk 11]
```diff
@@ -25,14 +26,81 @@ func (p *Plugin) PostTaskPreview(linkData []string, userID, channelID string) (*
...
+		},
+		Footer:     linkData[4],
+		FooterIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.ProjectIcon),
+	}
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
+	return post, ""
+}
+
+func (p *Plugin) PostPullRequestPreview(linkData []string, link, userID, channelID string) (*model.Post, string) {
+	pullRequest, _, err := p.Client.GetPullRequest(linkData[3], linkData[8], linkData[6], userID)
```
Comment (Kshitij-Katiyar) [inline server/plugin/taskLinkPreview.go:62]: is there a check for the length of linkData if there is nice if not please include this

File: server/plugin/taskLinkPreview_test.go [hunk 12]
```diff
@@ -20,15 +20,38 @@ func TestPostTaskPreview(t *testing.T) {
...
+		{
+			description: "PostPullRequestPreview: valid",
+			linkData:    []string{"https:", "", "dev.azure.com", "abc", "xyz", "_git", "xyz", "pullrequest", "1"},
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockedClient.EXPECT().GetPullRequest(gomock.Any(), gomock.Any(), gomock.Any(), "mockUserID").Return(&serializers.PullRequest{}, http.StatusOK, nil)
+			resp, err := p.PostPullRequestPreview(testCase.linkData, "mockPullRequestLink", "mockUserID", "mockChannelID")
+			assert.Equal(t, "", err)
+			assert.NotNil(t, resp)
```
Comment (Kshitij-Katiyar) [inline server/plugin/taskLinkPreview_test.go:54]: can we have a check for resp as well

File: server/plugin/taskLinkPreview_test.go [hunk 13]
```diff
@@ -20,15 +20,38 @@ func TestPostTaskPreview(t *testing.T) {
...
+		linkData    []string
+	}{
+		{
+			description: "PostPullRequestPreview: valid",
+			linkData:    []string{"https:", "", "dev.azure.com", "abc", "xyz", "_git", "xyz", "pullrequest", "1"},
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockedClient.EXPECT().GetPullRequest(gomock.Any(), gomock.Any(), gomock.Any(), "mockUserID").Return(&serializers.PullRequest{}, http.StatusOK, nil)
+			resp, err := p.PostPullRequestPreview(testCase.linkData, "mockPullRequestLink", "mockUserID", "mockChannelID")
```
Comment (Kshitij-Katiyar) [inline server/plugin/taskLinkPreview_test.go:52]:
```suggestion
			resp, stringErr := p.PostPullRequestPreview(testCase.linkData, "mockPullRequestLink", "mockUserID", "mockChannelID")
```

File: server/plugin/taskLinkPreview.go (hunk 11 above)
Comment (raghavaggarwal2308) [reply server/plugin/taskLinkPreview.go:62]: We don't have any direct check but we are using regex exp to check for the link and if the regex is passed then it definitely has this length

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/plugin_test.go [hunk 14]
```diff
@@ -76,52 +88,125 @@ func TestMessageWillBePosted(t *testing.T) {
...
+			regex:        constants.PullRequestLinkRegex,
+		},
+		{
+			description:  "IsLinkPresent: valid pull request link 4",
+			msg:          "\n\nhttp://dev.azure.com/abc/xyz/_git/xyz/pullrequest/1   mock-text",
+			expectedData: []string{"http:", "", "dev.azure.com", "abc", "xyz", "_git", "xyz", "pullrequest", "1"},
+			isValid:      true,
+			expectedLink: "http://dev.azure.com/abc/xyz/_git/xyz/pullrequest/1",
+			regex:        constants.PullRequestLinkRegex,
+		},
```
Comment (avas27JTG) [inline server/plugin/plugin_test.go:183]: this is good don't remove it now, but if you see ideally we don't need to check these many variations as there could be many possibilities of invalid cases here, so, 1 with valid and 1-2 for invalid would also be acceptable

File: webapp/src/styles/_components.scss [hunk 15]
```diff
@@ -59,3 +59,13 @@
...
+
+// Do not modify this class, this is a direct class from mattermost-webapp
+// This class is used for the dimensions of author_icon in Mattermost slack attachment
+.post .attachment {
+    // To make sure that this CSS does not conflict with the CSS of other plugins
+    img[src*="mattermost-plugin-azure-devops/static/repos-icon.svg"], img[src*="mattermost-plugin-azure-devops/static/boards-icon.svg"] {
+        width: 20px;
+        height: 20px;
+    }
+}
```
Comment (avas27JTG) [inline webapp/src/styles/_components.scss:71]: same this is also not synced with #87

Comment (raghavaggarwal2308) [reply webapp/src/styles/_components.scss:71]: I will take a pull when any one of the PR is merged

File: server/plugin/plugin_test.go (hunk 14 above)
Comment (raghavaggarwal2308) [reply server/plugin/plugin_test.go:183]: okay

File: server/constants/constants.go [hunk 16]
```diff
@@ -94,7 +93,13 @@ const (
...
 
+	// Colors
+	ReposIconColor  = "#d74f27"
+	BoardsIconColor = "#53bba1"
+
 	SubscriptionEventTypeDummy = "dummy"
 	GitBranchIcon              = "git-branch-icon.svg"
 	ProjectIcon                = "project-icon.svg"
+	ReposIcon                  = "repos-icon.svg"
+	BoardsIcon                 = "boards-icon.svg"
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/constants/constants.go:104]: will take a pull when one of the PR is merged

[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/constants/constants.go]: synced

File: webapp/src/styles/_components.scss (hunk 15 above)
Comment (raghavaggarwal2308) [reply webapp/src/styles/_components.scss]: synced

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**